### ISSUE TITLE
Losing annotations after analysis

### ISSUE THREAD

**Issue by Mevelo** [qupath/qupath#316]
**Describe the bug**
Unfortunately I am not able to share a lot of details (screenshots/reports) as it happenned a few days ago... but after all I thought it would be a good idea to share it here.
I was working on the analysis of some IF images. I created a project, set the images to fluorescence, applied the viewing settings I wanted, and started annotating the tissue. I have several annotations to draw on each tissue to mark specific areas, and it is relatively time consuming to do for 20 images. I also worked out the best cellular detections, trained a classifier, and got pretty much where I wanted.
I checked all images again, double checked all annotations and everything seemed fine.
I then started a batch analysis... which started normally. Once finished (and I didn't get any error message), I checked my detection and annotations results and >15 slides had no data. When I tried openning the images in the QuPath project, everything was gone: I get ask to set the image type and all annotations had disappear. I tried openning the .backup file but everything was gone there too. 
I'm thinking I must have done something wrong but I have no idea what... and unfortunately the .backup file was not enough to save me from doing everything again. I guess my questions are: do you know what could have caused this to happen? would there be a way of exporting the annotations separately (to save them for easy re-load if needed) ?

**Desktop:**
 - OS: macOS Sierra
 - QuPath Version 0.2

**Comment by petebankhead** [comment]
Oh dear, could it be this? https://github.com/qupath/qupath/issues/292

It seems that on Linux/Mac if .qpdata files are opened and saved externally (i.e. not from within a project) then they end up 'disappearing'. I don't think it should happen when everything is done within a project, but the outcome may differ depending up whether it is a project that was created with v0.1.2 or v0.2.0-m1/2.

**Comment by r-mvl** [comment]
Hm, perhaps.. In my case I worked on 0.2.0-m2 only, and within a project, so there must be more to it...

I've tried again with different images this morning after my post, and I found that processing 1 image worked, but running the analysis on a batch of 3 images did not... with a similar outcome (analysis not properly done and losing all annotations).
Except this time I did get an error message (maybe I missed it last time). Here is part of it:

```
INFO: 2159 nuclei detected (processing time: 21.18 seconds)
INFO: 2159 nuclei detected (10%)
INFO: 1803 nuclei detected (processing time: 19.79 seconds)
INFO: 1803 nuclei detected (11%)
ERROR: Error running plugin: java.lang.OutOfMemoryError: Java heap space
    at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
    at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
    at qupath.lib.plugins.AbstractPluginRunner.awaitCompletion(AbstractPluginRunner.java:237)
    at qupath.lib.plugins.AbstractPluginRunner.runTasks(AbstractPluginRunner.java:201)
    at qupath.lib.plugins.AbstractPlugin.runPlugin(AbstractPlugin.java:134)
    at qupath.lib.gui.scripting.QPEx.runPlugin(QPEx.java:258)
    at qupath.lib.gui.scripting.QPEx.runPlugin(QPEx.java:278)
    at qupath.lib.gui.scripting.QPEx$runPlugin.callStatic(Unknown Source)
    at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallStatic(CallSiteArray.java:55)
    at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callStatic(AbstractCallSite.java:196)
    at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callStatic(AbstractCallSite.java:216)
    at Script3.run(Script3.groovy:6)
    at org.codehaus.groovy.jsr223.GroovyScriptEngineImpl.eval(GroovyScriptEngineImpl.java:317)
    at org.codehaus.groovy.jsr223.GroovyScriptEngineImpl.eval(GroovyScriptEngineImpl.java:155)
    at qupath.lib.gui.scripting.DefaultScriptEditor.executeScript(DefaultScriptEditor.java:767)
    at qupath.lib.gui.scripting.DefaultScriptEditor.executeScript(DefaultScriptEditor.java:697)
    at qupath.lib.gui.scripting.DefaultScriptEditor.executeScript(DefaultScriptEditor.java:677)
    at qupath.lib.gui.scripting.DefaultScriptEditor$ProjectTask.call(DefaultScriptEditor.java:1288)
    at qupath.lib.gui.scripting.DefaultScriptEditor$ProjectTask.call(DefaultScriptEditor.java:1237)
    at javafx.concurrent.Task$TaskCallable.call(Task.java:1425)
    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
    at java.base/java.lang.Thread.run(Thread.java:834)
  Caused by Java heap space        at ij.process.FloatProcessor.snapshot(FloatProcessor.java:240)
        at ij.process.FloatProcessor.convolve(FloatProcessor.java:1069)
        at qupath.imagej.detect.cells.WatershedCellDetection$WatershedCellDetector.doDetection(WatershedCellDetection.java:600)
        at qupath.imagej.detect.cells.WatershedCellDetection$WatershedCellDetector.runDetection(WatershedCellDetection.java:997)
        at qupath.imagej.detect.cells.WatershedCellDetection$CellDetector.runDetection(WatershedCellDetection.java:362)
        at qupath.lib.plugins.DetectionPluginTools$DetectionRunnable.run(DetectionPluginTools.java:123)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
INFO: Completed with error java.lang.OutOfMemoryError: Java heap space
INFO: 
qupath.imagej.detect.cells.WatershedCellDetection  {"detectionImageFluorescence": 1,  "requestedPixelSizeMicrons": 0.1,  "backgroundRadiusMicrons": 0.0,  "medianRadiusMicrons": 0.0,  "sigmaMicrons": 0.9,  "minAreaMicrons": 6.0,  "maxAreaMicrons": 150.0,  "threshold": 2000.0,  "watershedPostProcess": true,  "cellExpansionMicrons": 3.0,  "includeNuclei": true,  "smoothBoundaries": true,  "makeMeasurements": true}
INFO: Training size: org.bytedeco.javacpp.opencv_core$Size[address=0x608000811620,position=0,limit=1,capacity=1,deallocator=org.bytedeco.javacpp.Pointer$NativeDeallocator[ownerAddress=0x608000811620,deallocatorAddress=0x13aaec9c0]]
INFO: Responses size: org.bytedeco.javacpp.opencv_core$Size[address=0x60800080d2a0,position=0,limit=1,capacity=1,deallocator=org.bytedeco.javacpp.Pointer$NativeDeallocator[ownerAddress=0x60800080d2a0,deallocatorAddress=0x13aaec9c0]]
INFO: RTrees classifier termination criteria: org.bytedeco.javacpp.opencv_core$TermCriteria[address=0x608000816130,position=0,limit=1,capacity=1,deallocator=org.bytedeco.javacpp.Pointer$NativeDeallocator[ownerAddress=0x608000816130,deallocatorAddress=0x13aae1520]]
ERROR: IOException exception reading file:/Volumes/Storage/Work/SLIDESCANS/190512_OLYMPUS_YKA_Batch/CRUK_YKA_16.1D_tam_2_20190513.vsi#1: x=9728, y=34816, w=512, h=512, z=0, t=0, downsample=1
    at java.base/java.nio.channels.spi.AbstractInterruptibleChannel.end(AbstractInterruptibleChannel.java:199)
    at java.base/sun.nio.ch.FileChannelImpl.endBlocking(FileChannelImpl.java:162)
    at java.base/sun.nio.ch.FileChannelImpl.readInternal(FileChannelImpl.java:816)
    at java.base/sun.nio.ch.FileChannelImpl.read(FileChannelImpl.java:796)
    at loci.common.NIOByteBufferProvider.allocateDirect(NIOByteBufferProvider.java:127)
    at loci.common.NIOByteBufferProvider.allocate(NIOByteBufferProvider.java:112)
    at loci.common.NIOFileHandle.buffer(NIOFileHandle.java:647)
    at loci.common.NIOFileHandle.<init>(NIOFileHandle.java:133)
    at loci.common.NIOFileHandle.<init>(NIOFileHandle.java:151)
    at loci.common.NIOFileHandle.<init>(NIOFileHandle.java:165)
    at loci.common.Location.getHandle(Location.java:522)
    at loci.common.Location.getHandle(Location.java:462)
    at loci.common.Location.getHandle(Location.java:443)
    at loci.common.Location.getHandle(Location.java:426)
    at loci.common.RandomAccessInputStream.<init>(RandomAccessInputStream.java:125)
    at loci.formats.in.CellSensReader.decodeTile(CellSensReader.java:1042)
    at loci.formats.in.CellSensReader.openBytes(CellSensReader.java:549)
    at loci.formats.FormatReader.openBytes(FormatReader.java:884)
    at loci.formats.ImageReader.openBytes(ImageReader.java:444)
    at loci.formats.ReaderWrapper.openBytes(ReaderWrapper.java:334)
    at loci.formats.ReaderWrapper.openBytes(ReaderWrapper.java:334)
    at loci.formats.gui.BufferedImageReader.openImage(BufferedImageReader.java:86)
    at qupath.lib.images.servers.bioformats.BioFormatsImageServer.readTile(BioFormatsImageServer.java:648)
    at qupath.lib.images.servers.AbstractTileableImageServer.getTile(AbstractTileableImageServer.java:61)
    at qupath.lib.images.servers.AbstractTileableImageServer.readBufferedImage(AbstractTileableImageServer.java:166)
    at qupath.lib.images.servers.AbstractTileableImageServer.readBufferedImage(AbstractTileableImageServer.java:19)
    at qupath.imagej.helpers.IJTools.convertToImagePlus(IJTools.java:536)
    at qupath.imagej.helpers.IJTools.convertToImagePlus(IJTools.java:573)
    at qupath.imagej.objects.PathImagePlus.getImage(PathImagePlus.java:156)
    at qupath.imagej.objects.PathImagePlus.getImage(PathImagePlus.java:120)
    at qupath.imagej.objects.PathImagePlus.getImage(PathImagePlus.java:47)
    at qupath.imagej.detect.cells.WatershedCellDetection$CellDetector.runDetection(WatershedCellDetection.java:269)
    at qupath.lib.plugins.DetectionPluginTools$DetectionRunnable.run(DetectionPluginTools.java:123)
    at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
    at java.base/java.lang.Thread.run(Thread.java:834)
ERROR: IOException exception reading file:/Volumes/Storage/Work/SLIDESCANS/190512_OLYMPUS_YKA_Batch/CRUK_YKA_16.1D_tam_2_20190513.vsi#1: x=9728, y=34816, w=512, h=512, z=0, t=0, downsample=1
    at java.base/java.nio.channels.spi.AbstractInterruptibleChannel.end(AbstractInterruptibleChannel.java:199)
    at java.base/sun.nio.ch.FileChannelImpl.endBlocking(FileChannelImpl.java:162)
    at java.base/sun.nio.ch.FileChannelImpl.readInternal(FileChannelImpl.java:816)
    at java.base/sun.nio.ch.FileChannelImpl.read(FileChannelImpl.java:796)
    at loci.common.NIOByteBufferProvider.allocateDirect(NIOByteBufferProvider.java:127)
    at loci.common.NIOByteBufferProvider.allocate(NIOByteBufferProvider.java:112)
    at loci.common.NIOFileHandle.buffer(NIOFileHandle.java:647)
    at loci.common.NIOFileHandle.<init>(NIOFileHandle.java:133)
    at loci.common.NIOFileHandle.<init>(NIOFileHandle.java:151)
    at loci.common.NIOFileHandle.<init>(NIOFileHandle.java:165)
    at loci.common.Location.getHandle(Location.java:522)
    at loci.common.Location.getHandle(Location.java:462)
    at loci.common.Location.getHandle(Location.java:443)
    at loci.common.Location.getHandle(Location.java:426)
    at loci.common.RandomAccessInputStream.<init>(RandomAccessInputStream.java:125)
    at loci.formats.in.CellSensReader.decodeTile(CellSensReader.java:1042)
    at loci.formats.in.CellSensReader.openBytes(CellSensReader.java:549)
    at loci.formats.FormatReader.openBytes(FormatReader.java:884)
    at loci.formats.ImageReader.openBytes(ImageReader.java:444)
    at loci.formats.ReaderWrapper.openBytes(ReaderWrapper.java:334)
    at loci.formats.ReaderWrapper.openBytes(ReaderWrapper.java:334)
    at loci.formats.gui.BufferedImageReader.openImage(BufferedImageReader.java:86)
    at qupath.lib.images.servers.bioformats.BioFormatsImageServer.readTile(BioFormatsImageServer.java:648)
    at qupath.lib.images.servers.AbstractTileableImageServer.getTile(AbstractTileableImageServer.java:61)
    at qupath.lib.images.servers.AbstractTileableImageServer.readBufferedImage(AbstractTileableImageServer.java:166)
    at qupath.lib.images.servers.AbstractTileableImageServer.readBufferedImage(AbstractTileableImageServer.java:19)
    at qupath.imagej.helpers.IJTools.convertToImagePlus(IJTools.java:536)
    at qupath.imagej.helpers.IJTools.convertToImagePlus(IJTools.java:573)
    at qupath.imagej.objects.PathImagePlus.getImage(PathImagePlus.java:156)
    at qupath.imagej.objects.PathImagePlus.getImage(PathImagePlus.java:120)
    at qupath.imagej.objects.PathImagePlus.getImage(PathImagePlus.java:47)
    at qupath.imagej.detect.cells.WatershedCellDetection$CellDetector.runDetection(WatershedCellDetection.java:269)
    at qupath.lib.plugins.DetectionPluginTools$DetectionRunnable.run(DetectionPluginTools.java:123)
    at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
    at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by petebankhead** [comment]
Ah, seems not to be the issue I linked to then... but rather a 'normal' (albeit very poorly handled) memory error.

Do you have more memory that you could make available to QuPath, or try reducing parallelization? 
There are some suggestions at https://github.com/qupath/qupath/wiki/Troubleshooting

Out-of-memory errors are (somewhat) inevitable when pushing things particularly hard, but if this results in completely losing .qpdata files then that definitely sounds like a bug that needs fixed. I will try to reproduce the problem and see if I can make it at least fail less catastrophically... and ideally not at all.

**Comment by r-mvl** [comment]
Fair enough. Thanks for looking in to this. It would be great if "at least" the annotations could be kept 👍 
Seeing the java heap space error I thought it might be that... but I have been able to run larger batches/images in the past so I wasn't expecting to have such problem. Can't remember if I actually did a larger analysis using this version though... and I don't get an error when not running it slide by slide (outside batch mode).

**Comment by petebankhead** [comment]
Yes, there may be an extra memory leak in there causing trouble...

I've also seen some Bio-Formats readers be rather more 'heavyweight' than others, meaning that if the image reading is parallelized this can be much more memory-hungry (since a separate reader is required per thread). In one case, the size of a .bfmemo file for a Zeiss .czi image revealed each reader was taking 1GB... and multichannel fluorescence images also need potentially much more memory to cache tiles suitably.

But whatever the cause, it shouldn't fail as badly as it does. How much memory is available to QuPath?

This may also be of some use: https://petebankhead.github.io/qupath/scripting/2018/03/06/script-memory-monitor.html

**Comment by r-mvl** [comment]
Sure, I see. It's true that I've recently started working with olympus .vsi files (as opposed to leica .scn) and it could be one of the factors.. Thanks for the script, very useful I will use it.

Currently I've set QuPath to 5gb (out of 8 on my laptop) and 3 processors.

**Comment by petebankhead** [comment]
Closing this because it has gone quiet and the issue could not be reproduced... hopefully the issue has been fixed along the path to v0.2.0 as a stable release.
